Themes: Make a ThemesSiteSelectorModal Higher-Order Component

ockham opened the pull request:
The idea here is to provide a higher-order component to just wrap `MultiSite` so that the latter can transparently use e.g. `props.options.activate` (just like `SingleSite`), with the HoC taking care of site selection and then invoking the correct action.

Furthermore, theme options are now declared at `connect` stage, which allows us to bind early (e.g. to site), again enabling us to just call the action in the wrapped component (without having to differentiate between binding or not binding there).

Turns out that as a side effect we can actually compose `MultiSite` and `SingleSite` from that HoC, and (un-`connect`ed) `LoggedOut`, so the latter is renamed to `ThemeShowcase`.

No visual changes, except for some label changes: _this PR now purports to change theme sheet preview action button functionality the way #6027 would do_. Check testing instructions there!

To test:
_This needs some thorough testing._ As this PR touches pretty much all theme actions, we need to test them all. 
* That is, every item inside the '...' menu for a random theme in logged-out and multi-site mode
* and also for an active one in single-site mode.
* Furthermore, try single site mode with a Jetpack site
* Finally, test the action button inside the Theme Preview
* And the 'Current Theme' bar options in single-site mode (again also with a Jetpack site).

TODO, in future PRs:
* Move `<Head>`  to `ThemeShowcase`, drop `makeElement`
* Pass `getScreenshotOption` to `ThemeSelection`, call there (or maybe only in `ThemesList`)
* Unify  `screenshotOption` and `defaultOption` handling
* Kill the `ThemeSelection` component and file, and inline their contents into `ThemeShowcase`. (That abstraction/indirection doesn't make much sense anymore.)
* Implement the `CurrentTheme` bar component in terms of theme `options`
* Rename `sourcePath` to `siteBasePath` for consistency with `my-sites/controller` and other site selector components
* Rename theme options to better reflect their labels. support -> setup, help -> support, preview -> demo etc.

Test live: https://calypso.live/?branch=update/themes-themes-site-selector-modal-hoc

cc @sirbrillig because, pie.

sirbrillig commented on client/my-sites/themes/single-site.jsx:
```diff
@@ -37,30 +37,31 @@ const ThemesSingleSite = ( props ) => {
-					illustration="/calypso/images/drake/drake-jetpack.svg" />
-			</Main>
-		);
+	if ( isJetpack ) {
```
Sorry to comment on in-progress work. Ignore this if it's not relevant. I think that this extra indentation is not actually easier to read. I think there's a guideline somewhere that goes into this in more detail, although I can't find it ATM. I suggest that if the repetition of `if ( isJetpack...` seems over-verbose, we use something like this instead, which could contain the whole of this block:

```js
if ( isJetPack ) {
  return this.renderJetpack( site );
}
...
renderJetpack( site ) {
  const jetpackEnabled = config.isEnabled( 'manage/themes-jetpack' );
  if ( ! jetpackEnabled ) {
    return ...
  }
  if ( ! site.hasJetpackThemes ) {
    return ...
  }
  if ( ! site.canManage ) {
  }
}
```

ockham replied:
Huh right, thanks for flattening those conditionals, should've thought of that myself.

Part of this has now gone into 9b89e9d. I haven't yet added a separate method for this since this is currently a stateless functional component -- tell me if you think it reads well enough now or if I should turn it into a normal component :-)

sirbrillig replied:
> since this is currently a stateless functional component

Oh right! I didn't see that before. Well, it's already one level of indentation better. I think my eyes start to hurt after 3. I'll say that's fine with me for now to keep it from having to be an object.

I haven't used functional components for more than three-line components before. This is an interesting exercise in how to compress complicated logic into small spaces. I suppose you could define additional render methods as separate functions within the closure, but that feels maybe a little awkward? Are there any patterns I could learn here?

sirbrillig replied:
Hm, maybe a pattern would be something like using another functional component, which could even be defined within the same file since it wouldn't be used outside the closure.

```js
if ( isJetpack ) {
  return <JetpackThemePage site={ site } />
}
```

ockham replied:
Good call! 1f3677d

ockham commented:
Okay, I'd say this is ready for review.

sirbrillig commented on client/my-sites/themes/logged-out.jsx:
```diff
@@ -1,108 +1,58 @@
+		{
+			options: merge(
+				{},
+				{
```
I really want to move this object to somewhere above here to reduce indentation and also just clarify the three functions in `connect`, but of course we need `dispatchProps` from the closure. In addition, it'd be nice to make it clear what this object does. Perhaps if there was a generator function for the options object that had defaults? I'm still considering this in my head. Maybe you'll have additional ideas.

ockham replied:
I'm leaning towards just defining the whole closure as

```js
const mergeProps = ( stateProps, dispatchProps, ownProps ) => Object.assign( /* ... */ )
```

since we'll want to use (part of) that object elsewhere anyway (for the sheet).

sirbrillig commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -0,0 +1,117 @@
+	},
+
+	onPreviewButtonClick( theme ) {
+		const defaultOption = this.props.options[ this.props.defaultOption ];
```
`defaultOption` isn't a required prop, so this statement could produce null. We should probably either check for the function before calling it, require the prop (and its action), or have a default prop with a noop action.

sirbrillig commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -0,0 +1,117 @@
+			[ 'preview', 'action' ],
+			theme => this.togglePreview( theme )
+		),
+			buttonOptions = merge(
```
can we make each of these a separate `const` statement? I think it's more clear than comma-separated and more consistent with other calypso code I've seen.

ockham replied:
Oh yeah, it was really rather ugly like that. 63b4e77

sirbrillig commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -0,0 +1,117 @@
+					action: previewAction
+				} }
+			),
+			defaultOption = this.props.options[ this.props.defaultOption ],
```
same comment as above: what happens if this statement produces `null`?

sirbrillig commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -0,0 +1,117 @@
+				} }
+			),
+			defaultOption = this.props.options[ this.props.defaultOption ],
+			getScreenshotOption = theme => buttonOptions[ this.props.getScreenshotOption( theme ) ];
```
similar to the above, `getScreenshotOption` isn't required, and even it it were, the reference could produce `null` if the object property doesn't exist.

sirbrillig commented:
This is looking really great! Some comments above, but I need to dig and test a bit more too.

sirbrillig commented:
I guess the thing I keep thinking about the most is: if all the action variations can be derived from the state, why have the `options` object at all? Why not just compute the possible actions in the component? I'm sure it's more complicated than that, but that's my thoughts right now. I'll probably change my mind later. 😉

ockham commented:
> I guess the thing I keep thinking about the most is: if all the action variations can be derived from the state, why have the `options` object at all? Why not just compute the possible actions in the component? I'm sure it's more complicated than that, but that's my thoughts right now. I'll probably change my mind later. 😉

The main reason for not doing it inside the component is that the way I'm doing it with this PR keeps the component more options-agnostic than it could otherwise be. I've added two examples to the (updated) PR description, but maybe they were out of context and not as concise, so I'll try to rephrase them here:
`ThemeShowcase` can always just call `option.action()` (let's assume `option` is `activate` for example), and the correct thing will happen -- so in single-site mode, it will activate the theme for the given site; while in multi-site mode, the `ThemesSiteSelectionModal` has intercepted and decorated that `option`, so it will fire up the site selector modal, and only after the user has picked a site will it activate the theme on that site. I'll argue that's better than before in terms of separation of concerns -- `ThemeShowcase` can be entirely ignorant of that modal and all the state handling that it requires.

Compare that to the pre-PR approach where each component (`LoggedOut`, `MultiSite`, and `SingleSite`) would duplicate most functionality, with only the way `options` were defined really differing.

I agree that the `options` objects still are kinda ugly. There's a couple of points where we can trim down their ugliness, but I'd rather do that in separate PRs to not unnecessarily inflate this one even more.
* Replacing helpers by actual selectors (#6165) will make the binding to `site` for `single-site` at `connect` stage even more compelling, I believe.
* Some actions don't really depend on context but are always there (the ones pointing to the theme sheet, i.e. Info, Setup, and Support). I think we'll be able to move those to a place where they'll be automatically shared by all three modes/contexts (logged-out, multi-site, single-site).

Finally, I do want to declare those options (and the logic that selects which ones are relevant in a given context) outside a Themes List specific component in order to be able to reuse them for the theme sheet button. I guess the theme sheet button option might be equivalent to either `options[ defaultOption ]` -- or maybe we'll have to define `sheetButtonOption`.

`</Filibuster>`

sirbrillig commented:
That's a great explanation! Thank you for taking the time to lay it out for me. In that case, I agree with this approach. It's certainly less complex than what's there.

sirbrillig commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -14,7 +14,38 @@ import Main from 'components/main';
+	addTracking
+} from './helpers';
+
+export const sheetOptions = ( site = false, isJetpack = false ) => ( {
```
❤️ it! Great idea. Super tiny suggestion: give the function name a verb, like `getSheetOptions()`

ockham commented:
I'm considering making each option a function (since most use a `theme` param for `action` or `getUrl` anyway). This will allow more consistency with the dreaded `getScreenshotOption`. Maybe it'll end up  inside `options` after all, alongside `defaultOption`.

Roughly like so:

```js
/* ... */
const signup = theme => ( {
	label: i18n.translate( 'Pick this design', {
		comment: 'when signing up for a WordPress.com account with a selected theme'
	} ),
	action: dispatchProps.signup.bind( theme ),
	url: getSignupUrl( theme ),
} );

options = {
 /* ...*/
		signup,
		info,
		default: signup
		screenshot: theme => theme.someFlag ? signup( theme ) : info ( theme )
}
```

That's probably going to be defaultOption/screenshotOption fix, part 2. Tomorrow :-)

ockham commented:
Hmm, experimented a bit, and it seems that changing individual options to functions would be a bit too intrusive for the scope of this PR. Going to postpone that to a later one, and just guard against cases https://github.com/Automattic/wp-calypso/pull/6349#discussion_r69173949 for now.

ockham commented:
Note to self -- the changes to `ThemeSiteSelectorModal` might have broken the theme sheet!

ockham commented:
My inner perfectionist would like to tweak the `connect()` statements some more to make them even terser. The pragmatist feels refactoring fatigue creeping in. 

The scope of this PR is big enough -- making it work with the sheet required some more rather substantial changes -- so I'm setting this back to 'Needs Review'. I've also updated the PR description a little. Note that I've now changed the sheet default actions to (hopefully) emulate #6027.

folletto commented:
Tested on Chrome:

1. Logged out + Theme Sheet + Pick this design ✅
2. Logged out + Theme Sheet + Live Demo + Pick this design ✅
3. Logged out + ••• + Pick this design ✅
4. Logged out + ••• + Live demo + Pick this design ✅
5. All Sites + Theme Sheet + Activate + Select Site + Activate + Welcome ✅
6. All Sites + Theme Sheet + Live Demo + Activate + Select Site + Activate + Welcome ✅
7. All Sites + ••• + Activate + Select Site + Welcome ✅
8. All Sites + ••• + Live Demo + Try & Customize + Select Site + Save and Activate ✅ 
9. All Sites + ••• + Try & Customize + Select Site ✅
10. Single Site + Theme Sheet + Activate + Welcome ✅
11. Single Site + Theme Sheet + Live Demo + Activate + <del>Welcome</del> Customize? ❌
12. Single Site + ••• + Activate + Welcome ✅
13. Single Sites + ••• + Live Demo + Try & Customize + Save and Activate ✅
14. Single Sites + ••• + Try & Customize ✅
15. Jetpack + ••• + Info ✅
16. Jetpack + ••• + Try & Customize ✅
17. Jetpack + ••• + Live demo ✅  _(same as Try?)_
18. Single Site + Current Bar + Customize ✅
19. Single Site + Current Bar + Info ✅
20. Jetpack + Current Bar + Customize ✅
21. Jetpack + Current Bar + Info ✅

I hope I've covered all the instances ^^'

Note that (11) above works, but in practice the WebPreview doesn't get closed, thus the Welcome message isn't visible. The button simply changes from "Activate" to "Customize". Not sure if it's an issue of this PR tho.

ockham commented:
> Note that (11) above works, but in practice the WebPreview doesn't get closed, thus the Welcome message isn't visible. The button simply changes from "Activate" to "Customize". Not sure if it's an issue of this PR tho.

Probably not in particular newly introduced but just not covered before (where 'Try and Custmize' would just open the customizer, thus making the preview go away). I'll see how much hassle it'd be to fix this.

seear commented on client/my-sites/theme/main.jsx:
```diff
@@ -80,36 +88,12 @@ const ThemeSheet = React.createClass( {
+		this.props.defaultOption.action( this.props );
 	},
 
 	onPreviewButtonClick() {
```
`onPrimaryClick()` and `onPreviewButtonClick()` are now the same and could be combined into one function.

ockham replied:
14d32b6550068db85d109632cb764069f3199bf0

seear commented on client/my-sites/theme/main.jsx:
```diff
@@ -268,7 +252,7 @@ const ThemeSheet = React.createClass( {
 
 	renderPreview() {
-		const buttonLabel = this.props.isLoggedIn ? i18n.translate( 'Try & Customize' ) : i18n.translate( 'Pick this design' );
+		const buttonLabel = this.props.defaultOption.label;
```
We can lose this local now if we want.

ockham replied:
9e8091c7c98ebccb12fe31daa1daeee8667ba0c2

seear commented on client/my-sites/theme/main.jsx:
```diff
@@ -387,12 +353,57 @@ const ThemeSheet = React.createClass( {
 } );
 
+const WrappedThemeSheet = ( props ) => {
+	if ( ! props.isLoggedIn || props.selectedSite ) {
```
Could maybe eliminate this branch by having an `enabled` prop on `<ThemesSiteSelectorModal/>`. Fine as is, though.

ockham replied:
I considered turning `<ThemesSiteSelectorModal/>` into a more multi-purpose `<ThemesActionBinder/>` that would bind single-site actions to `site` and just pass through its children when logged-out, but that seemed to dilute concerns. I'd rather let consumers of the component handle if they use it or not. Ideally tho this could be expressed more tersely, by a ternary and some sort of identity/pass-thru component...

I guess I'd rather leave as-is for now :-)

sirbrillig commented:
Looks great to me! ❤️ that logged-out.js. So small!

ockham commented:
@folletto a3883ef should fix (11)

folletto commented:
(11) Single Site + Theme Sheet + Live Demo + Activate + Welcome ✅

Confirmed 👍

seear commented:
Something odd, can't see how it would relate to this PR but I don't see it on master:
* turn on mobile emulation in chrome and visit http://calypso.localhost:3000/design

<img width="1085" alt="screen shot 2016-07-13 at 18 20 11" src="https://cloud.githubusercontent.com/assets/7767559/16813035/bfdc6c8a-4927-11e6-820d-1d71da922d68.png">

Maybe just needs a rebase.

seear commented:
Same problem with mobile on the live branch.

ockham commented:
Rebased -- seems to fix it. (Note -- could only repro with Chrome set to emulate iPhone, not S5.)

seear commented on client/my-sites/themes/themes-site-selector-modal.jsx:
```diff
@@ -14,49 +17,98 @@ import { trackClick } from './helpers';
+	},
+
+	isThemeOrActionSet() {
+		return ! isEmpty( this.state.selectedTheme ) || ! isEmpty( this.state.selectedOption );
```
So if the action is set, but not the theme, we should go ahead and show the preview?

ockham replied:
Do you mean this should be `isThemeAndActionSet()`, and a `&&`? Not quite sure right now why we originally decided on the `||`...

ockham replied:
> show the preview?

The site selector modal you mean, right?

seear replied:
> The site selector modal you mean, right?

Yes sorry. Let's not hold up the PR for this, since it has been working fine.

seear commented on client/my-sites/themes/themes-site-selector-modal.jsx:
```diff
@@ -14,49 +17,98 @@ import { trackClick } from './helpers';
+			{},
+			defaultOption,
+			defaultOption.action
+				? { action: theme => this.showSiteSelectorModal( defaultOption, theme ) }
```
`wrappedOptions` and `wrappedDefaultOption` might be able to share some code here? (fine to iterate in another PR)

ockham replied:
https://github.com/Automattic/wp-calypso/pull/6762

seear commented:
ok looks good :shipit: 😄
